src/openms/include/OpenMS/DATASTRUCTURES/QTCluster.h

```diff
@@ -110,124 +109,134 @@ namespace OpenMS
 */
   class OPENMS_DLLAPI QTCluster
   {
-private:
+public:
 
     // need to store more than one
-    typedef std::multimap<double, GridFeature*> NeighborListType;
-    typedef OpenMSBoost::unordered_map<Size, NeighborListType> NeighborMapMulti;
-
-    typedef std::pair<double, GridFeature*> NeighborPairType;
-    typedef OpenMSBoost::unordered_map<Size, NeighborPairType> NeighborMap;
-
-    /// Pointer to the cluster center
-    GridFeature* center_point_;
-
-    /**
-     * @brief Map that keeps track of the best current feature for each map
-     *
-     */
-    NeighborMap neighbors_;
-
-    /**
-     * @brief Temporary map tracking *all* neighbors
-     *
-     * For each input run, a multimap which contains pointers to all
-     * neighboring elements and the respective distance.
-     *
-     */
-    NeighborMapMulti* tmp_neighbors_;
-
-    /// Maximum distance of a point that can still belong to the cluster
-    double max_distance_;
+    typedef std::multimap<double, const GridFeature*> NeighborList;
+    typedef OpenMSBoost::unordered_map<Size, NeighborList> NeighborMapMulti;
 
-    /// Number of input maps
-    Size num_maps_;
-
-    /// Quality of the cluster
-    double quality_;
+    struct Neighbor
+    {
+      double distance;
+      const GridFeature* feature;
+    };
 
-    /// Has the cluster changed (if yes, quality needs to be recomputed)?
-    bool changed_;
+    typedef OpenMSBoost::unordered_map<Size, Neighbor> NeighborMap;
 
-    /// Keep track of peptide IDs and use them for matching?
-    bool use_IDs_;
+    struct Element
+    {
+      Size map_index;
+      const GridFeature* feature;
+    };
 
-    /// Whether current cluster is valid
-    bool valid_;
+    typedef std::vector<Element> Elements;
 
     /** 
-     * @brief Whether initial collection of all neighbors is needed
-     *
-     * This variable stores whether we need to collect all annotations first
-     * before we can decide upon the best set of cluster points. This is
-     * usually only necessary if the center point does not have an annotation
-     * but we want to use ids.
-     *
-    */
-    bool collect_annotations_;
-
-    /// Whether current cluster is accepting new elements or not (if true, no more new elements allowed)
-    bool finalized_;
-
-    /// x coordinate in the grid cell
-    Int x_coord_;
-
-    /// y coordinate in the grid cell
-    Int y_coord_;
+     * @brief Class to store the bulk internal data (neighbors, annotations, etc.)
+     * 
+     * Has no functionality without a QTCluster pointing to it.
+     * Create object of this class before calling constructor of QTCluster
+     */
+    class BulkData
```

Comment:
do these internal classes and structs need to be public?

Reply:
Currently it has to be, because we want to store the `BulkData` objects in a vector and therefore construct them independently of the `QTCluster`objects. We hope to improve cache behavior by this. Canging it will probably lead to a small performance penalty. Only some seconds though. Would you prefer it if we changed it?

Reply:
hmm, not sure. depends a bit on if this would give users an "easier" (e.g., fewer parameters passed) interface and if it hides some of the added complexity. I let you or @cbielow decide.

Reply:
We decided to stick with the performance. 

One could implement a wrapper that hides the ugly interface and presents a better one. But this would probably result in yet another pretty big change in this code and would not be worth the effort in my opinion.